### PR TITLE
DM: self-distillation via EMA teacher consistency loss

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The EMA model (decay=0.9995) is already computed every step and used only for best-checkpoint selection. We hypothesize that its smoothed predictions carry a richer, lower-variance teaching signal than the raw ground-truth labels alone. By adding a consistency regularization term that nudges the online model toward the EMA model's predictions, we can extract a free training signal — effectively implementing Born Again Neural Networks (Furlanello et al., 2018, https://arxiv.org/abs/1805.04770) within a single training run.

Loss formulation:
```
loss = (1 - alpha) * MSE(pred, target) + alpha * MSE(pred, ema_pred)
```

Where `ema_pred` is computed with `torch.no_grad()` from the EMA model (no gradient flows through the teacher). This adds zero parameters and negligible compute.

Two trials:
- **Trial A**: alpha=0.3 (stronger distillation signal)
- **Trial B**: alpha=0.1 (lighter regularization, closer to baseline)

## Instructions

Start from the DM champion config (PR #3072). Apply the following changes:

**1. Smoke test first** (1–2 epochs, both alpha values) to confirm loss decreases and no NaN/divergence before committing to full runs.

**2. Implementation — add to `train.py`:**

In the training loop, after computing the standard supervised loss, compute the EMA model's prediction on the same batch and add the consistency term:

```python
# After: loss = criterion(pred, target)
with torch.no_grad():
    ema_pred = ema_model(x)  # use whatever the EMA model forward call is
distill_loss = F.mse_loss(pred, ema_pred)
loss = (1 - alpha) * loss + alpha * distill_loss
```

Log `distill_loss` separately in W&B every step so we can track how the teacher-student gap evolves over training.

**3. Run both trials with the champion DM config:**

**Trial A (alpha=0.3):**
```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml \
    --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
    --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
    --epochs 999 \
    --batch-size 1 \
    --drivaerml-train-surface-points 50000 \
    --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --ema-decay 0.9995 \
    --self-distill-alpha 0.3 \
    --wandb_group dm-self-distillation
```

**Trial B (alpha=0.1):**
```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml \
    --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
    --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
    --epochs 999 \
    --batch-size 1 \
    --drivaerml-train-surface-points 50000 \
    --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --ema-decay 0.9995 \
    --self-distill-alpha 0.1 \
    --wandb_group dm-self-distillation
```

If `--self-distill-alpha` is not yet a CLI flag, implement it and add it. The default should be 0.0 (disabled) so existing runs are unaffected.

**4. Report back with:**
- Best `val_primary/surface_rel_l2_pct` for each trial (from the EMA checkpoint, not the online model)
- W&B run IDs for both trials
- A plot or logged values of `distill_loss` over training to show teacher-student gap evolution
- Whether alpha=0.3 or alpha=0.1 performed better and any intuition why

## Baseline

Current DrivAerML best (target to beat):
- **val** `surface_rel_l2_pct`: **3.833%** (epoch 511)
- **test** `surface_rel_l2_pct`: **4.685%**
- Best PR: #3072 (eren — EMA=0.9995 + gc=0.5, 4L/512d/8H, AdamW lr=5e-4, T_max=30)
- W&B run: `ncl1dh88` (eren/dm-ema-9995-gc05)
- Reproduce command:
```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml \
    --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
    --batch-size 1 --drivaerml-train-surface-points 50000 \
    --drivaerml-eval-surface-points 50000 --max-train-batches 394 \
    --max-eval-batches 200 --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT canute:

## Results

### Summary

Both trials failed to beat the baseline. Both eventually diverged due to the self-distillation consistency loss causing late-stage training instability. Stronger alpha diverged earlier.

| Metric | Baseline (#3072) | Trial A (α=0.3) | Trial B (α=0.1) |
|--------|-------------------|------------------|------------------|
| **val `surface_rel_l2_pct`** | **3.833%** (ep 511) | 6.446% (ep 159) | 4.323% (ep 335) |
| **test `surface_rel_l2_pct`** | **4.685%** | 7.167% | 5.252% |
| Diverged at | — | ~epoch 170 | ~epoch 400 |

### Trial A — alpha=0.3 (stronger distillation)

- **Best val**: 6.446% at epoch 159
- **Best test**: 7.167% (from best checkpoint)
- **W&B run**: `1df67nu8` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/1df67nu8))
- **Outcome**: Diverged around epoch 170. Grad norms hit Infinity and val regressed to ~61%. Never recovered. The strong distillation signal (30% of loss) destabilized training once the EMA teacher started lagging behind the online model.

Command:
```bash
cd target/icml2026
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml \
    --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
    --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
    --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --ema-decay 0.9995 --self-distill-alpha 0.3 \
    --wandb-group dm-self-distillation --wandb-name "canute/dm-self-distill-alpha03" \
    --agent canute
```

### Trial B — alpha=0.1 (lighter regularization)

- **Best val**: 4.323% at epoch 335
- **Best test**: 5.252% (from best checkpoint)
- **W&B run**: `r8c2ztmm` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/r8c2ztmm))
- **Outcome**: Performed better and survived longer, but also diverged around epoch 400. Grad norms hit Infinity, val regressed to ~51%. The best val (4.32%) was closer to baseline (3.83%) but never beat it — and the instability cut short what might have been further convergence.

Command:
```bash
cd target/icml2026
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml \
    --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
    --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
    --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --ema-decay 0.9995 --self-distill-alpha 0.1 \
    --wandb-group dm-self-distillation --wandb-name "canute/dm-self-distill-alpha01" \
    --agent canute
```

### Distill loss evolution

The `distill_loss` metric was logged per epoch in both W&B runs. Key observations:
- **Early training** (epochs 1-50): distill_loss starts at ~0.06 and declines to ~0.01 as the EMA teacher converges toward the online model
- **Mid training** (epochs 50-150): stabilizes around 0.003-0.01 — the teacher and student are close
- **Pre-divergence**: distill_loss begins spiking, indicating the EMA teacher is providing conflicting signals as the online model's gradients destabilize
- **Post-divergence**: the pattern depends on which diverged first — in Trial A, distill_loss collapsed to near-zero as both models plateaued at the same bad solution

### What happened

The self-distillation hypothesis is sound in principle, but the implementation exposed a critical failure mode: **late-stage training instability**. The EMA teacher (decay=0.9995) lags the online model by ~2000 steps of effective averaging. In later training, when the model is near convergence and gradients are small, the distillation term can push the online model toward a slightly stale teacher prediction, creating a feedback loop that amplifies small perturbations. The stronger the alpha, the earlier this instability manifests:
- α=0.3 → diverged ~epoch 170
- α=0.1 → diverged ~epoch 400

The grad_clip at 0.5 was insufficient to prevent divergence once the feedback loop started — grad norms went to Infinity.

### Suggested follow-ups

1. **Alpha warmup + decay schedule**: Ramp alpha from 0 to target over first 20% of training, hold, then decay back to 0 in the final 10-20%. This avoids both the "noisy teacher" problem in early training and the "stale teacher" instability in late training. (Supported by "Weak-to-Strong KD" — arXiv:2604.15451)
2. **Much smaller alpha** (0.01-0.05): The instability is proportional to alpha. Very small alpha might give a mild regularization benefit without destabilizing.
3. **Stop-gradient with EMA checkpoint restoration**: When divergence is detected (e.g., val metric regresses 2x from best), restore the best EMA checkpoint and disable distillation for the remainder of training.
4. **Separate EMA for distillation**: Use a lower EMA decay (e.g., 0.999) for the distillation teacher so it tracks the online model more closely, reducing the teacher-student lag.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3225 because ADVISOR: Closing — self-distillation via EMA teacher is a dead end. Alpha=0.3: val=6.446% (+68%), diverged ep170. Alpha=0.1: val=4.323% (+12.8%), diverged ep400. Structural flaw: EMA=0.9995 creates ~2000-step teacher lag that becomes a destabilizing feedback loop in late training — the distillation term overcomes gc=0.5 and sends grad norms to infinity. The monotonic relationship (higher alpha → earlier divergence) confirms the mechanism. Clean extrapolation suggests very small alpha (0.01-0.03) might survive longer but would provide negligible signal. Your analysis was excellent — particularly identifying the feedback loop mechanism..
